microcosm_eventsource/controllers.py

```diff
@@ -0,0 +1,134 @@
+
+        """
+        self.validate_required_fields(event_type, **kwargs)
+        parent = self.store.retrieve_most_recent(**kwargs)
+        version, state = self.process_state_transition(parent, event_type)
+        event = self.create_event(parent, event_type, version, state, **kwargs)
+        self.publish_event(event)
+        return event
+
+    def validate_required_fields(self, event_type, **kwargs):
+        """
+        Validate type-specific required fields.
+
+        """
+        missing_required_fields = [
+            required_field
+            for required_field in event_type.value.requires
+            if kwargs.get(required_field) is None
+        ]
+        if missing_required_fields:
+            raise with_context(
+                UnprocessableEntity("Validation error"), [
+                    {
+                        "message": "Missing required field: '{}'".format(
+                            camelize(required_field, uppercase_first_letter=False),
+                        ),
+                        "field": camelize(required_field, uppercase_first_letter=False),
+                        "reasons": [
+                            "Event type '{}' requires '{}'".format(
+                                event_type.name,
+                                camelize(required_field, uppercase_first_letter=False),
+                            )
+                        ],
+                    }
+                    for required_field in missing_required_fields
+                ],
+            )
+
+    def process_state_transition(self, parent, event_type):
+        """
+        Process a state transition.
+
+        Determines whether a state transition is legal, what version to use, and what state
+        to persist.
+
+        """
+        if not event_type.is_legal_after(parent):
+            if parent is None:
+                # event may not be initial
+                raise IllegalInitialStateError("Event type '{}' may not be initial".format(
+                    event_type.name,
+                ))
+            else:
+                # event may not follow previous
+                raise IllegalStateTransitionError("Event type '{}' may not follow [{}]".format(
+                    event_type.name,
+                    ", ".join(item.name for item in parent.state),
+                ))
+
+        parent_version = parent.version if parent else 1
+        parent_state = parent.state if parent else []
+
+        if event_type.is_accumulating:
+            state = parent_state + [event_type]
+        else:
+            state = [event_type]
+
+        if event_type.is_restarting:
+            version = parent_version + 1
+        else:
+            version = parent_version
+
+        return version, state
+
+    def create_event(self, parent, event_type, version, state, **kwargs):
+        """
+        Create the event.
+
+        """
+        # XXX we should upsert here as long as we can reason about the constraints
```

Comment:
There are race conditions we can solve with upsert, but it requires a bit more care with our unique constraints.